indexing-service/src/main/java/io/druid/indexing/overlord/supervisor/SupervisorResource.java

```diff
@@ -0,0 +1,230 @@
+            }
+
+            return Response.ok(spec.get()).build();
+          }
+        }
+    );
+  }
+
+  @POST
+  @Path("/{id}/shutdown")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response shutdown(@PathParam("id") final String id)
+  {
+    return asLeaderWithSupervisorManager(
+        new Function<SupervisorManager, Response>()
+        {
+          @Override
+          public Response apply(SupervisorManager manager)
+          {
+            if (!manager.hasSupervisor(id)) {
+              return Response.status(Response.Status.NOT_FOUND)
+                             .entity(ImmutableMap.of("error", String.format("[%s] does not exist", id)))
+                             .build();
+            }
+
+            manager.stopAndRemoveSupervisor(id);
+            return Response.ok(ImmutableMap.of("id", id)).build();
+          }
+        }
+    );
+  }
+
+  @GET
+  @Path("/history")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response specGetAllHistory()
+  {
+    return asLeaderWithSupervisorManager(
+        new Function<SupervisorManager, Response>()
+        {
+          @Override
+          public Response apply(SupervisorManager manager)
+          {
+            return Response.ok(manager.getSupervisorHistory()).build();
+          }
+        }
+    );
+  }
+
+  @GET
+  @Path("/{id}/history")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response specGetHistory(@PathParam("id") final String id)
+  {
+    return asLeaderWithSupervisorManager(
+        new Function<SupervisorManager, Response>()
+        {
+          @Override
+          public Response apply(SupervisorManager manager)
+          {
+            Map<String, List<VersionedSupervisorSpec>> history = manager.getSupervisorHistory();
+            if (history.containsKey(id)) {
+              return Response.ok(history.get(id)).build();
+            } else {
+              return Response.status(Response.Status.NOT_FOUND)
+                             .entity(
+                                 ImmutableMap.of(
+                                     "error",
+                                     String.format(
+                                         "No history for [%s] (history available for %s)",
+                                         id,
+                                         history.keySet()
+                                     )
+                                 )
+                             )
+                             .build();
+            }
+          }
+        }
+    );
+  }
+
+  private Response asLeaderWithSupervisorManager(Function<SupervisorManager, Response> f)
+  {
+    Optional<SupervisorManager> supervisorManager = taskMaster.getSupervisorManager();
+    if (supervisorManager.isPresent()) {
+      return f.apply(supervisorManager.get());
+    } else {
+      // Encourage client to try again soon, when we'll likely have a redirect set up
+      return Response.status(Response.Status.SERVICE_UNAVAILABLE).build();
```

Comment:
yes, there should be redirection in place already, this is really a SERVER_ERROR scenario, request shouldn't land here in general if this overlord is not leader.